libraries/standard/http/src/http_client.c

```diff
@@ -245,6 +245,816 @@ static int findHeaderValueParserCallback( http_parser * pHttpParser,
+     * affect future parsing as the changed segment will always be before the
+     * next place to parse. */
+    /* coverity[misra_c_2012_rule_11_8_violation] */
+    pNextWriteLoc = ( uint8_t * ) ( pResponse->pBody + pResponse->bodyLen );
+
+    /* If the response is of type Transfer-Encoding: chunked, then actual body
+     * will follow the the chunked header. This body data is in a later location
+     * and must be moved up in the buffer. When pLoc is greater than the current
+     * end of the body, that signals the parser found a chunk header. */
+    if( pLoc > ( const char * ) pNextWriteLoc )
+    {
+        ( void ) memcpy( pNextWriteLoc, ( const uint8_t * ) pLoc, length );
+    }
+
+    /* Increase the length of the body found. */
+    pResponse->bodyLen += length;
+
+    /* Set the next location of parsing. */
+    pParsingContext->pBufferCur = pLoc + length;
+
+    LogDebug( ( "Response parsing: Found the response body. "
+                "BodyLength=%d",
+                length,
+                pLoc ) );
+
+    return shouldContinueParse;
+}
+
+/*-----------------------------------------------------------*/
+
+static int httpParserOnMessageCompleteCallback( http_parser * pHttpParser )
+{
+    HTTPParsingContext_t * pParsingContext = NULL;
+
+    assert( pHttpParser != NULL );
+    assert( pHttpParser->data != NULL );
+
+    pParsingContext = ( HTTPParsingContext_t * ) ( pHttpParser->data );
+
+    /* The response message is complete. */
+    pParsingContext->state = HTTP_PARSING_COMPLETE;
+
+    LogDebug( ( "Response parsing: Response message complete." ) );
+
+    return HTTP_PARSER_CONTINUE_PARSING;
+}
+
+/*-----------------------------------------------------------*/
+
+static void initializeParsingContextForFirstResponse( HTTPParsingContext_t * pParsingContext )
+{
+    assert( pParsingContext != NULL );
+
+    /* Initialize the third-party HTTP parser to parse responses. */
+    http_parser_init( &( pParsingContext->httpParser ), HTTP_RESPONSE );
+
+    http_parser_set_max_header_size( HTTP_MAX_RESPONSE_HEADERS_SIZE_BYTES );
+
+    /* No response has been parsed yet. */
+    pParsingContext->state = HTTP_PARSING_NONE;
+
+    /* No response to update is associated with this parsing context yet. */
+    pParsingContext->pResponse = NULL;
+}
+
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t processHttpParserError( http_parser * pHttpParser )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    assert( pHttpParser != NULL );
+
+    switch( ( enum http_errno ) ( pHttpParser->http_errno ) )
+    {
+        case HPE_OK:
+            /* There were no errors. */
+            break;
+
+        case HPE_INVALID_EOF_STATE:
+
+            /* In this case the parser was passed a length of zero, which indicates
+             * an EOF from the server, in the middle of parsing the response.
+             * This case is already handled by checking HTTPParsingContext_t.state. */
+            break;
```

Comment:
That is correct. However, the behavior is expected because `shouldRecv` will become false since `parsingContext.state != HTTP_PARSING_COMPLETE`. The appropriate error will be returned by `getFinalResponseStatus(...)`.